src/Control/Monad/Freer/Interpretation.hs

```diff
@@ -0,0 +1,213 @@
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE MonoLocalBinds        #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Control.Monad.Freer.Interpretation where
```

Comment:
This module has the interpreter combinators. I renamed them so their names are a little more meaningful to humans; I figured since the types have breaking changes anyway we might as well.

Reply:
“Meaningful” is subjective. :) My experience is that some of the original names are more meaningful to a certain set of people, but perhaps less useful to others more familiar with other concepts. Here was my general philosophy for `freer-simple`’s names:

- A library ought to introduce as few concepts and as few terms of art as possible as long as they are entirely internally-consistent.
- When possible, terms should be borrowed from English words or phrases so as to borrow intuition via analogy.
- However, if an analogy is likely to break down quickly, especially if it is in confusing ways, then it’s more harmful than helpful. A fresh, artificial term should be used instead so as to avoid potentially-confusing baggage.

The metaphor I used for `freer-simple` is one that already exists in computer science: languages and interpreters. This metaphor is so ubiquitous that it appears in Gang of Four as the *interpreter pattern*, so it seems safe to assume any prior exposure is consistent with my use of the terms. In this analogy, effects are languages, and effect handlers are interpreters.

These are the only *fundamental* concepts in freer-simple’s vocabulary. Therefore, everything else really ought to fall out of application of the metaphor to the derived concepts. Let’s evaluate how well that’s worked, shall we?

- We’ll start with the simplest handler combinator, `interpret`. This name is essentially one of our axioms, so obviously it’s consistent with the metaphor.

- Next, let’s look at `translate`. I believe this to be a good name, since it fits into the metaphor very well. The process of moving from one language to another while preserving meaning is *translation*; that’s what this function does, so its name is good.

- What about `reinterpret` and friends? These names are a little less good upon close examination, but I think the metaphor still holds up. Morally, they are very much like `translate`, but since they run in a context in which all the other effects are accessible, they also have the ability to do some extra interpretation.

   In practice, `reinterpret` is far more useful than `translate`, since `translate` demands you use the effect constructors directly (e.g. `Ask` or `Get`), rather than using the wrapping functions (e.g. `ask` or `get`). Since effects are treated like abstractions, thinking about the constructors might be unnecessarily low-level. It’s a subtle distinction, one that I don’t expect most newcomers to immediately understand. Perhaps `translate` ought to be somehow deemphasized for that reason.

- `interpose` exists somewhat outside of the interpreter metaphor, instead living more inside the client/server metaphor that `handleRelay` uses. Perhaps a better name would be something like `embellish`, but honestly, that seems too cute to me, and the existing name still does make some sense: it’s interposing on an effect’s interpretation by some other interpreter.

- `subsume` needs little explanation, since it just does what it says. This name needs no metaphors!

- Finally, there are the “variant” handlers. `interpretM` interprets an effect monadically, so the name works. `interpretWith` and `interposeWith` are generalized versions of `interpret` and `interpose` that expose the continuation. The names don’t really connect to what they do (what does `With` have to do with being given the continuation?), so cleverer names would be better, but these are also the “advanced” handlers, which I don’t expect to be used by very many people, anyway.

All together, these names form what I think is a mostly-coherent linguistic abstraction. Ideally, a user of `freer-simple` should be thinking about an “effect” as a sort of primitive thing. In my ideal world, I would be able to make defining an effect use the syntax

```haskell
effect State s where
  get :: s
  put :: s -> ()
```

but I can’t, so I have to kludge it a bit. In any case, I’m really angling to make the abstraction as watertight as possible, without any strings attached.

---

Alright, so, what does that small essay have to do with your PR? Well, I want to give the context under which I’m evaluating your proposed changes. You propose a new function, `natural`, as well as a new name for `translate`, `replace`, but these don’t fit in with the metaphor at all. `natural` even demands that users be familiar with an unrelated concept, natural transformations. `replace` is vague and nonspecific (the name can come up *everywhere* in programming), and `natural` isn’t even a verb. The names aren’t *bad*, not in isolation, but they’re certainly inconsistent with the rest of the API.

I can make similar arguments for things like `relay`, `transform`, and `introduce`. The names aren’t *bad*, but they don’t obey any consistent metaphor. Is `relay` using the client/server metaphor? Are `replace` and `introduce` using a data structure metaphor (specifically, a stack)? `transform` is a bit better, but it sure has a hell of a type for something with that simple a name!

`stateful` is better, since it’s a name with an immediately obvious meaning, but I don’t actually understand why it needs to exist at all. It could be done just as easily using `reinterpret` combined with `runState`. It seems to run counter to the idea that most people are going to be defining effects in terms of existing ones, anyway. (The name is also not a verb.)

To summarize a bit, and to put things more bluntly, I don’t like these name changes, but I’m also not dead set against them. Obviously, you found the original names less helpful than I intended, so just because *I* think they follow a cohesive metaphor doesn’t they do. On the other hand, I’ve gotten a lot of very positive feedback from users of `freer-simple`’s API who have said the metaphor has been helpful to them, so I’d rather not throw it all out without some better understanding about what your philosophy is behind your proposed API.

Reply:
Thanks for your thoughts! TBH I implemented these in a greenfield, and so they're not renamed so much as "the first things I came up with." Your point about `stateful` is well taken---mine performs better than the two passes, but this could just as easily be a rewrite rule.

Reply:
I don't have a particularly strong horse in this race, but would like to respond to a few points:

While a metaphor might be useful, it requires me to buy into it. Something I hate about `pipes` and `conduits` eg. is that in order to use them I need to read a really big tutorial to understand what the heck they're trying to talk about. In comparison, `streaming` doesn't require a metaphor, it's just "here are things you already know, but they're streaming."

Given the `streaming` point of view, `natural` is a good name _because_ it accurately describes what it does, assuming I already know what a natural transformation is. `transform` likewise lifts the effects of a monad transformer directly into an interpretation.

While I don't often agree with Ed Kmett's documentation strategy, one thing I do like is his inclination to "call something by its name." Given that natural transformations and monad transformers are particularly useful ideas _outside_ of this domain, maintaining that clarity seems valuable to me.

Reply:
I understand. I think it’s probably more or less impossible to know if the metaphor is actually a net positive or net negative without doing some kind of rigorous study involving actual people, and that’s obviously not going to happen anytime soon, so we’re stuck with anecdotal evidence and personal feelings. I think the point of view you give is quite reasonable and has merits; I even agree with it a lot of the time.

But: given I don’t know whether the existing nomenclature is good or not, and I don’t think either of us can provide much evidence that something different would be meaningfully superior to the “newbie Haskeller” (since obviously neither of us are that), my preference is to avoid changing things up. I’d like to stick with the names we have now.